onnxruntime/python/tools/transformers/convert_generation.py

```diff
@@ -1075,37 +1075,73 @@ def update_decoder_subgraph_past_present_share_buffer(subg):
+    subg: GraphProto, is_beam_search: bool, switch_attention: bool
+) -> bool:
     """Update the Attention nodes to DecoderMaskedMultiheadAttention.
 
     Args:
         subg (GraphProto): GraphProto of the decoder subgraph
+        is_beam_search (bool): Boolean specifying if the sampling algo is BeamSearch
+        switch_attention (bool): Boolean specifying if `Attention` is to be switched with `DecoderMaskedMultiheadAttention`
     """
-    decoder_masked_attention_supported_attr = [
-        "past_present_share_buffer",
-        "num_heads",
-        "scale",
-        "domain",
-    ]
-    new_nodes = []
-    for node in subg.node:
-        if node.op_type == "Attention":
-            kwargs = kwargs_of(node)
-            for k in kwargs.copy():
-                # The Attention operator does not support different qkv hidden sizes when past/present
-                # input/output exists (GPT2 model). Hence, we should never run into this.
-                # But, if we do, do not go ahead with the optimization.
-                if k == "qkv_hidden_sizes":
-                    return False
-
-                if k not in decoder_masked_attention_supported_attr:
-                    del kwargs[k]
-            nis = []
-            nis.extend(node.input)
-            node = onnx.helper.make_node("DecoderMaskedMultiheadAttention", nis, node.output, name=node.name, **kwargs)
-        new_nodes.extend([node])
-    subg.ClearField("node")
-    subg.node.extend(new_nodes)
+    if is_beam_search:
+        new_inputs = []
+        for i, vi in enumerate(subg.input):
+            new_inputs.extend([vi])
+
+        # Add 2 BeamSearch specific inputs
+        new_inputs.extend([onnx.helper.make_tensor_value_info("beam_width", onnx.TensorProto.INT32, shape=[1])])
+        new_inputs.extend(
+            [
+                onnx.helper.make_tensor_value_info(
+                    "cache_indirection", onnx.TensorProto.INT32, shape=["batch_size", "beam_width", "max_seq_len"]
+                )
+            ]
+        )
+        subg.ClearField("input")
+        subg.input.extend(new_inputs)
+
+    if switch_attention:
+        decoder_masked_attention_supported_attr = [
+            "past_present_share_buffer",
+            "num_heads",
+            "scale",
+            "domain",
+        ]
+
+        new_nodes = []
+        for node in subg.node:
+            if node.op_type == "Attention":
+                kwargs = kwargs_of(node)
+                for k in kwargs.copy():
+                    # The Attention operator does not support different qkv hidden sizes when past/present
+                    # input/output exists (GPT2 model). Hence, we should never run into this.
+                    # But, if we do, do not go ahead with the optimization.
+                    if k == "qkv_hidden_sizes":
+                        return False
+
+                    if k not in decoder_masked_attention_supported_attr:
+                        del kwargs[k]
```

Comment:
Add some logging of warning: remove attribute (except unidirectional) could cause problems.

Reply:
Good idea, I meant to do this based on your comment in the other PR but somehow it slipped through.

Reply:
Added warning log while removing any attribute except "unidirectional"